File: server/lib/translate.js
"use strict";
//
// LocalBox server — перевод внутренних сообщений (entity-store, Ecast-стиль) в события Blobcast
// (старый протокол): RoomBlobChanged, CustomerJoinedRoom, CustomerMessage и т.п.
// Собственная реализация LocalBox (описывает лишь формат событий Blobcast).
//
// ctxUserId — userId, относящийся к событию (для JoinRoom — входящий игрок; для CustomerMessage —
// отправитель; для CustomerLeftRoom — ушедший). Пустой объект = нечего слать этому клиенту.
//

module.exports = function translate(msg, roomId, ctxUserId, hostUserId) {
    const r = msg.result || {};
    switch (msg.opcode) {
        case "client/connected":
            if (r.role !== "player") return {};
            return {
                type: "Event", event: r.reconnect ? "CustomerRejoinedRoom" : "CustomerJoinedRoom",
                roomId, customerUserId: r.userId, customerName: r.name,
                options: { roomcode: "", name: r.name, email: "", phone: "" },
            };
        case "client/disconnected":
        case "client/kicked":
            if (r.role !== "player") return {};
            return { type: "Event", event: "CustomerLeftRoom", roomId, customerUserId: ctxUserId };
        case "client/send":
            return { type: "Event", event: "CustomerMessage", roomId, userId: ctxUserId, message: r.body };
        case "client/welcome":
            if (ctxUserId === hostUserId) return {}; // хосту JoinRoom не нужен
            return {
                type: "Result", action: "JoinRoom", success: true, initial: !r.reconnect,
                roomId, joinType: r.profile ? "player" : "audience", userId: ctxUserId,
                options: { roomcode: "", name: r.name, email: "", phone: "" },
            };
        case "room/exit":
            return { type: "Event", event: "RoomDestroyed", roomId };
        case "object":
            if (r.key && r.key.indexOf("bc:") === 0)
                return { type: "Event", event: r.key === "bc:room" ? "RoomBlobChanged" : "CustomerBlobChanged", roomId, blob: r.val };
            return {};
        default:
            return {};
    }
};

File: android/server/translate.go
package main

//
// LocalBox Go-сервер — перевод внутренних сообщений (entity-store, Ecast) в события Blobcast
// (старый протокол). Порт server/lib/translate.js. Пустая карта = слать нечего.
//

import "strings"

func translate(msg map[string]interface{}, roomID, ctxUserID, hostUserID string) map[string]interface{} {
	r, _ := msg["result"].(map[string]interface{})
	if r == nil {
		r = map[string]interface{}{}
	}
	opcode, _ := msg["opcode"].(string)

	switch opcode {
	case "client/connected":
		if str(r["role"]) != "player" {
			return nil
		}
		event := "CustomerJoinedRoom"
		if toBool(r["reconnect"]) {
			event = "CustomerRejoinedRoom"
		}
		return map[string]interface{}{
			"type": "Event", "event": event, "roomId": roomID,
			"customerUserId": str(r["userId"]), "customerName": str(r["name"]),
			"options": map[string]interface{}{"roomcode": "", "name": str(r["name"]), "email": "", "phone": ""},
		}
	case "client/disconnected", "client/kicked":
		if str(r["role"]) != "player" {
			return nil
		}
		return map[string]interface{}{"type": "Event", "event": "CustomerLeftRoom", "roomId": roomID, "customerUserId": ctxUserID}
	case "client/send":
		return map[string]interface{}{"type": "Event", "event": "CustomerMessage", "roomId": roomID, "userId": ctxUserID, "message": r["body"]}
	case "client/welcome":
		if ctxUserID == hostUserID {
			return nil // хосту JoinRoom не нужен
		}
		joinType := "audience"
		if r["profile"] != nil {
			joinType = "player"
		}
		return map[string]interface{}{
			"type": "Result", "action": "JoinRoom", "success": true, "initial": !toBool(r["reconnect"]),
			"roomId": roomID, "joinType": joinType, "userId": ctxUserID,
			"options": map[string]interface{}{"roomcode": "", "name": str(r["name"]), "email": "", "phone": ""},
		}
	case "room/exit":
		return map[string]interface{}{"type": "Event", "event": "RoomDestroyed", "roomId": roomID}
	case "object":
		key := str(r["key"])
		if strings.HasPrefix(key, "bc:") {
			event := "CustomerBlobChanged"
			if key == "bc:room" {
				event = "RoomBlobChanged"
			}
			return map[string]interface{}{"type": "Event", "event": event, "roomId": roomID, "blob": r["val"]}
		}
		return nil
	default:
		return nil
	}
}
